dataset-processor-python/data_loader.py

```diff
@@ -31,25 +43,186 @@ class DatasetWriter(DataLoader):
+            # if not number, return the original value
+            return value
+
+    def process_features(self, column, value, features):
+        if self.is_valid_feature_value(value):
+            features.append(value)
+            self.update_feature_metadata(column, value)
+        else:
+            print(f"Invalid value: {value} in column {column}")
+            return
+
+    def is_valid_feature_value(self, value):
+        return pd.isna(value) or isinstance(value, (int, float))
+
+    def update_feature_metadata(self, column, value):
+        is_discrete = self.is_discrete(value)
+        self.discrete_features_dict[column] = is_discrete
+        if column not in self.feature_def_keys:
+            self.feature_def_keys.append(column)
+
+    def is_discrete(self, numeric_value):
         """
-        create a folder with the database_name
-        initialize the folder with required json files
+        Determines if the numeric value is discrete
         """
-        # folder_name = self.dataset_name + "_" + version
-        # path = os.path.join("data", folder_name)
-        # os.makedirs(path, exist_ok=True)
-        # create json files -> feature_defs.json, dataset.json, cell_feature_analysis.json, image_settings.json
-        pass
+        return not pd.isna(numeric_value) and numeric_value == int(numeric_value)
+
+    def compile_cell_feature_analysis(self, file_info, features):
+        self.cell_feature_analysis_file.append(
+            {"file_info": file_info, "features": features}
+        )
+
+    def compile_feature_defs(self):
+        description = ""
+        tooltip = ""
+        for key in self.feature_def_keys:
+            discrete = self.discrete_features_dict[key]
+            unit = self.get_unit(key)
+            stripped_key = key.replace(f"({unit})", "").strip()
+            self.features_data_order.append(stripped_key)
+            display_name = DatasetWriter.format_display_name(
+                stripped_key.replace("-", " ")
+            )
+            feature_def = {
+                "key": stripped_key,
+                "displayName": display_name,
+                "unit": unit,
+                "description": description,
+                "tooltip": tooltip,
+                "discrete": discrete,
+            }
+            if discrete:
+                column_data = self.get_column_data(key)
+                options = self.write_discrete_feature_options(column_data)
+                feature_def["options"] = options
+            self.feature_defs_file.append(feature_def)
+
+    @staticmethod
+    def format_display_name(title):
+        lowercase_list = ["the", "for", "in", "of", "on", "to", "and", "as", "or"]
+        return " ".join(
+            word if word in lowercase_list else word.title() for word in title.split()
+        )
+
+    @staticmethod
+    def get_unit(key):
+        """
+        Extracts the unit from the key if present
+        """
+        match = re.search(r"\((.*?)\)", key)
```

Comment:
I think this regex might not work for units with nested parentheses? For example, if you have `feature (unit(mm))` it looks like it would only select `unit(mm` and miss the second closing parentheses.

Reply:
The feature titles we get from the csv, like `interphase-or-mitosis (stage)`, are matched using a regex that works with single parentheses. But you're right, if there are more complex formats, we can adjust the regex accordingly